File: backend/gods_eye/detection/live_stream_urls.py
"""Resolve common «IP Camera» iOS/Android (ShenYao) home URLs to MJPEG stream candidates."""

from __future__ import annotations

from urllib.parse import urlparse, urlunparse


def live_http_candidate_urls(stream_url: str) -> list[str]:
    raw = (stream_url or "").strip()
    if not raw:
        return []
    p = urlparse(raw)
    if p.scheme not in ("http", "https"):
        return [raw]

    path = (p.path or "/").rstrip("/") or "/"
    has_real_path = path != "/"

    ordered: list[str] = []
    seen: set[str] = set()

    def add(u: str) -> None:
        u = u.strip()
        if u and u not in seen:
            seen.add(u)
            ordered.append(u)

    add(raw)

    if not has_real_path and not p.query:
        origin = urlunparse((p.scheme, p.netloc, "", "", "", ""))
        for suffix in ("/video", "/videofeed", "/video.mjpg", "/mjpeg"):
            add(origin + suffix)

    return ordered

File: frontend/gods_eye/src/app/core/utils/live-stream-urls.ts
/**
 * 《IP Camera》/ IP Camera for iOS (ShenYao) and similar apps often show a *home* URL
 * like `http://192.168.1.5:8081/` while the actual MJPEG stream is on a subpath.
 * Browsers and OpenCV need the MJPEG URL, not the HTML landing page.
 */
export function liveHttpCandidateUrls(streamUrl: string): string[] {
  const raw = streamUrl.trim();
  if (!raw) {
    return [];
  }
  if (!/^https?:\/\//i.test(raw)) {
    return [raw];
  }
  let parsed: URL;
  try {
    parsed = new URL(raw);
  } catch {
    return [raw];
  }
  if (parsed.protocol !== 'http:' && parsed.protocol !== 'https:') {
    return [raw];
  }

  const path = (parsed.pathname || '/').replace(/\/+$/, '') || '/';
  const hasRealPath = path !== '/' && path.length > 1;

  const origin = `${parsed.protocol}//${parsed.host}`;

  const ordered = new Set<string>();
  ordered.add(raw);

  if (!hasRealPath && !parsed.search) {
    // ShenYao IP Camera (iOS/Android) common MJPEG paths — try after exact URL fails
    const suffixes = ['/video', '/videofeed', '/video.mjpg', '/mjpeg', '/?action=stream'];
    for (const s of suffixes) {
      ordered.add(`${origin}${s}`);
    }
  }

  return [...ordered];
}
